election-server/src/bin/indexer.rs

```diff
@@ -0,0 +1,580 @@
+                Ok(time) => {
+                    successive_db_errors = 0;
+                    tracing::info!(
+                        "Processed block {} at height {} transactions in {}ms",
+                        block_data.block_hash,
+                        block_data.height.height,
+                        time.num_milliseconds()
+                    );
+                    retry_block_data = None;
+                }
+                Err(e) => {
+                    successive_db_errors += 1;
+                    // wait for 2^(min(successive_errors - 1, 7)) seconds before attempting.
+                    // The reason for the min is that we bound the time between reconnects.
+                    let delay = std::time::Duration::from_millis(
+                        500 * (1 << std::cmp::min(successive_db_errors, 8)),
+                    );
+                    tracing::warn!(
+                        "Database connection lost due to {:#}. Will attempt to reconnect in {}ms.",
+                        e,
+                        delay.as_millis()
+                    );
+                    tokio::time::sleep(delay).await;
+
+                    // Get new db connection from the pool
+                    db = match db_pool
+                        .get()
+                        .await
+                        .context("Failed to get new database connection from pool")
+                    {
+                        Ok(db) => db,
+                        Err(e) => {
+                            block_receiver.close();
+                            return Err(e);
+                        }
+                    };
+                    retry_block_data = Some(block_data);
+                }
+            }
+        } else {
+            break;
+        }
+    }
+
+    block_receiver.close();
+
+    Ok(())
+}
+
+/// Inserts the `block_data` collected for a single block into the database
+/// defined by `db`. Everything is commited as a single transactions allowing
+/// for easy restoration from the last recorded block (by height) inserted into
+/// the database. Returns the duration it took to process the block.
+#[tracing::instrument(skip(db))]
+async fn db_insert_block<'a>(
+    db: &mut Database,
+    block_data: &'a BlockData,
+) -> anyhow::Result<chrono::Duration> {
+    let start = chrono::Utc::now();
+    let transaction = db
+        .client
+        .transaction()
+        .await
+        .context("Failed to build DB transaction")?;
+
+    {
+        let transaction = Transaction::from(&transaction);
+        transaction.set_latest_height(block_data.height).await?;
+
+        for ballot in block_data.ballots.iter() {
+            transaction
+                .insert_ballot(ballot, block_data.block_time)
+                .await?;
+        }
+    }
+
+    let now = tokio::time::Instant::now();
+    transaction
+        .commit()
+        .await
+        .context("Failed to commit DB transaction.")?;
```

Comment:
If the reason for the "double reference" in the Transaction wrapper is that you need the inner transaction here, I suggest just having the field public and accessing the commit method.

Reply:
This was the reason 😄 well spotted, detective!